pkg/kubefnord/util/util.go

```diff
@@ -0,0 +1,216 @@
+type fedConfig struct {
+	pathOptions   *clientcmd.PathOptions
+	hostConfig    *rest.Config
+	clusterConfig *rest.Config
+	fedConfig     *rest.Config
+}
+
+// NewFedConfig creates a fedConfig for `kubefnord` commands.
+func NewFedConfig(pathOptions *clientcmd.PathOptions) FedConfig {
+	return &fedConfig{
+		pathOptions: pathOptions,
+	}
+}
+
+// PathOptions provides filesystem based kubeconfig access.
+func (a *fedConfig) PathOptions() *clientcmd.PathOptions {
+	return a.pathOptions
+}
+
+// NewHostConfig provides a rest config to talk to the host kubernetes cluster
+// based on the context and kubeconfig passed in.
+func (a *fedConfig) NewHostConfig(context, kubeconfigPath string) (*rest.Config, error) {
+	hostConfig := a.getClientConfig(context, kubeconfigPath)
+	hostClientConfig, err := hostConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	a.hostConfig = hostClientConfig
+	return a.hostConfig, nil
+}
+
+// HostConfig retrieves the host kubernetes cluster config.
+func (a *fedConfig) HostConfig() *rest.Config {
+	return a.hostConfig
+}
+
+// NewFedConfig provides a rest config to talk to the federation cluster
+// based on the context and kubeconfig passed in.
+func (a *fedConfig) NewFedConfig(context, kubeconfigPath string) (*rest.Config, error) {
+	fedConfig := a.getClientConfig(context, kubeconfigPath)
+	fedClientConfig, err := fedConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	a.fedConfig = fedClientConfig
+	return a.fedConfig, nil
+}
+
+// FedConfig retrieves the host kubernetes cluster config.
+func (a *fedConfig) FedConfig() *rest.Config {
+	return a.fedConfig
+}
+
+// NewClusterConfig provides a rest config to talk to the joining kubernetes
+// cluster based on the context and kubeconfig passed in.
+func (a *fedConfig) NewClusterConfig(context, kubeconfigPath string) (*rest.Config, error) {
+	clusterConfig := a.getClientConfig(context, kubeconfigPath)
+	clusterClientConfig, err := clusterConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	a.clusterConfig = clusterClientConfig
+	return a.clusterConfig, nil
+}
+
+// ClientConfig retrieves the joining kubernetes cluster config.
+func (a *fedConfig) ClusterConfig() *rest.Config {
+	return a.clusterConfig
+}
+
+// getClientConfig is a helper method to create a client config from the
+// context and kubeconfig passed as arguments.
+func (a *fedConfig) getClientConfig(context, kubeconfigPath string) clientcmd.ClientConfig {
+	loadingRules := *a.pathOptions.LoadingRules
+	loadingRules.Precedence = a.pathOptions.GetLoadingPrecedence()
+	loadingRules.ExplicitPath = kubeconfigPath
+	overrides := &clientcmd.ConfigOverrides{
+		CurrentContext: context,
+	}
+
+	return clientcmd.NewNonInteractiveDeferredLoadingClientConfig(&loadingRules, overrides)
+}
+
+// FedClientset provides the clients to talk to the kubernetes
+// API server, the cluster registry API server, and the federation API server.
+type FedClientset interface {
```

Comment:
There is potential for a common interface like this to be shared across tests and controllers.